### ISSUE TITLE
Parallel pod deletion makes kubectl stuck forever

### ISSUE THREAD

**Issue by ilia1243** [kubernetes/kubernetes#123177]
### What happened?

`kubectl delete pod` may become stuck during execution of the following steps:
* Restart kubelet `systemctl restart kubelet`
* Force remove all containers for static pods `crictl ps --name '(kube-apiserver|kube-scheduler|kube-controller-manager|etcd)' -q | xargs -I CONTAINER sudo crictl rm -f CONTAINER`
* Try to delete some static pod `kubectl delete pod -n kube-system kube-controller-manager-k8s-control-plane-1`

Result: kubectl prints that the pod is deleted, but never exits.

It seems that the problem is caused by parallel deletion of the pod by kubelet.
Please see details in the steps to reproduce.

### What did you expect to happen?

kubectl exits.

### How can we reproduce it (as minimally and precisely as possible)?

1. Install single-node cluster
2. Run script
<details>

```sh
#!/bin/bash

for i in {1..100} ; do
  echo "Attempt $i"
  systemctl restart kubelet
  crictl ps --name '(kube-apiserver|kube-scheduler|kube-controller-manager|etcd)' -q | xargs -I CONTAINER sudo crictl rm -f CONTAINER
  for component in kube-apiserver kube-scheduler kube-controller-manager etcd; do
    while true ; do
      echo "$(date '+%H:%M:%S,%N') delete $component"
      if kubectl --v=10 delete pod -n kube-system ${component}-k8s-control-plane-1 2> failed.txt ; then
        break
      fi
    done
  done
  for component in kube-apiserver kube-scheduler kube-controller-manager etcd; do
    while true ; do
      echo "$(date '+%H:%M:%S,%N') get $component"
      if kubectl get pod -n kube-system ${component}-k8s-control-plane-1 | grep '1/1' ; then
        break
      fi
      sleep 5
    done
  done
done
```
</details>

Example logs when kubectl became stuck during deletion of `kube-controller-manager`
<details>

```console
18:57:25,065592823 delete kube-apiserver
pod "kube-apiserver-k8s-control-plane-1" deleted
18:57:27,788489678 delete kube-scheduler
pod "kube-scheduler-k8s-control-plane-1" deleted
18:57:29,219558884 delete kube-controller-manager
pod "kube-controller-manager-k8s-control-plane-1" deleted
```
</details>

Snippet from kube-apiserver audit logs
<details>

```console
{"kind":"Event","apiVersion":"audit.k8s.io/v1","level":"Metadata","auditID":"90c1d55b-5937-4f09-9526-4aeee3580438","stage":"ResponseComplete","requestURI":"/api/v1/namespaces/kube-system/pods/kube-controller-manager-k8s-control-plane-1","verb":"get","user":{"username":"system:node:k8s-control-plane-1","groups":["system:nodes","system:authenticated"]},"sourceIPs":["<IP>"],"userAgent":"kubelet/v1.28.4 (linux/amd64) kubernetes/bae2c62","objectRef":{"resource":"pods","namespace":"kube-system","name":"kube-controller-manager-k8s-control-plane-1","apiVersion":"v1"},"responseStatus":{"metadata":{},"code":200},"requestReceivedTimestamp":"2024-02-07T18:57:27.243300Z","stageTimestamp":"2024-02-07T18:57:27.248216Z","annotations":{"authorization.k8s.io/decision":"allow","authorization.k8s.io/reason":""}}
{"kind":"Event","apiVersion":"audit.k8s.io/v1","level":"Metadata","auditID":"8fb73098-9b08-4415-a358-c64b8c3380da","stage":"ResponseComplete","requestURI":"/api/v1/namespaces/kube-system/pods/kube-controller-manager-k8s-control-plane-1","verb":"delete","user":{"username":"kubernetes-admin","groups":["system:masters","system:authenticated"]},"sourceIPs":["<IP>"],"userAgent":"kubectl/v1.28.4 (linux/amd64) kubernetes/bae2c62","objectRef":{"resource":"pods","namespace":"kube-system","name":"kube-controller-manager-k8s-control-plane-1","apiVersion":"v1"},"responseStatus":{"metadata":{},"code":200},"requestReceivedTimestamp":"2024-02-07T18:57:29.329926Z","stageTimestamp":"2024-02-07T18:57:29.351432Z","annotations":{"authorization.k8s.io/decision":"allow","authorization.k8s.io/reason":""}}
{"kind":"Event","apiVersion":"audit.k8s.io/v1","level":"Metadata","auditID":"49641f1d-63b1-40a1-92ed-679bba85321c","stage":"ResponseComplete","requestURI":"/api/v1/namespaces/kube-system/pods/kube-controller-manager-k8s-control-plane-1","verb":"get","user":{"username":"kubernetes-admin","groups":["system:masters","system:authenticated"]},"sourceIPs":["<IP>"],"userAgent":"kubectl/v1.28.4 (linux/amd64) kubernetes/bae2c62","objectRef":{"resource":"pods","namespace":"kube-system","name":"kube-controller-manager-k8s-control-plane-1","apiVersion":"v1"},"responseStatus":{"metadata":{},"code":200},"requestReceivedTimestamp":"2024-02-07T18:57:29.353780Z","stageTimestamp":"2024-02-07T18:57:29.370011Z","annotations":{"authorization.k8s.io/decision":"allow","authorization.k8s.io/reason":""}}
{"kind":"Event","apiVersion":"audit.k8s.io/v1","level":"Metadata","auditID":"e7b68471-b675-492e-901a-dbc17c13f704","stage":"ResponseComplete","requestURI":"/api/v1/namespaces/kube-system/pods/kube-controller-manager-k8s-control-plane-1","verb":"delete","user":{"username":"system:node:k8s-control-plane-1","groups":["system:nodes","system:authenticated"]},"sourceIPs":["<IP>"],"userAgent":"kubelet/v1.28.4 (linux/amd64) kubernetes/bae2c62","objectRef":{"resource":"pods","namespace":"kube-system","name":"kube-controller-manager-k8s-control-plane-1","apiVersion":"v1"},"responseStatus":{"metadata":{},"code":200},"requestReceivedTimestamp":"2024-02-07T18:57:29.344031Z","stageTimestamp":"2024-02-07T18:57:29.383082Z","annotations":{"authorization.k8s.io/decision":"allow","authorization.k8s.io/reason":""}}
{"kind":"Event","apiVersion":"audit.k8s.io/v1","level":"Metadata","auditID":"dceca62f-7d6a-4245-921a-82ef4f9e57fd","stage":"ResponseComplete","requestURI":"/api/v1/namespaces/kube-system/pods?fieldSelector=metadata.name%3Dkube-controller-manager-k8s-control-plane-1\u0026limit=500\u0026resourceVersion=0","verb":"list","user":{"username":"kubernetes-admin","groups":["system:masters","system:authenticated"]},"sourceIPs":["<IP>"],"userAgent":"kubectl/v1.28.4 (linux/amd64) kubernetes/bae2c62","objectRef":{"resource":"pods","namespace":"kube-system","name":"kube-controller-manager-k8s-control-plane-1","apiVersion":"v1"},"responseStatus":{"metadata":{},"code":200},"requestReceivedTimestamp":"2024-02-07T18:57:29.371784Z","stageTimestamp":"2024-02-07T18:57:29.384505Z","annotations":{"authorization.k8s.io/decision":"allow","authorization.k8s.io/reason":""}}
{"kind":"Event","apiVersion":"audit.k8s.io/v1","level":"Metadata","auditID":"672ade25-6311-4d48-a46b-62df25430918","stage":"ResponseStarted","requestURI":"/api/v1/namespaces/kube-system/pods?allowWatchBookmarks=true\u0026fieldSelector=metadata.name%3Dkube-controller-manager-k8s-control-plane-1\u0026resourceVersion=84421\u0026timeoutSeconds=307\u0026watch=true","verb":"watch","user":{"username":"kubernetes-admin","groups":["system:masters","system:authenticated"]},"sourceIPs":["<IP>"],"userAgent":"kubectl/v1.28.4 (linux/amd64) kubernetes/bae2c62","objectRef":{"resource":"pods","namespace":"kube-system","name":"kube-controller-manager-k8s-control-plane-1","apiVersion":"v1"},"responseStatus":{"metadata":{},"code":200},"requestReceivedTimestamp":"2024-02-07T18:57:29.385687Z","stageTimestamp":"2024-02-07T18:57:29.386369Z","annotations":{"authorization.k8s.io/decision":"allow","authorization.k8s.io/reason":""}}
{"kind":"Event","apiVersion":"audit.k8s.io/v1","level":"Metadata","auditID":"f7a6d0fd-16ce-495d-9d69-c00d12bbd1fc","stage":"ResponseComplete","requestURI":"/api/v1/namespaces/kube-system/pods","verb":"create","user":{"username":"system:node:k8s-control-plane-1","groups":["system:nodes","system:authenticated"]},"sourceIPs":["<IP>"],"userAgent":"kubelet/v1.28.4 (linux/amd64) kubernetes/bae2c62","objectRef":{"resource":"pods","namespace":"kube-system","name":"kube-controller-manager-k8s-control-plane-1","uid":"0cc56e74843e8ff67c46e61b93c89605","apiVersion":"v1"},"responseStatus":{"metadata":{},"code":201},"requestReceivedTimestamp":"2024-02-07T18:57:29.385166Z","stageTimestamp":"2024-02-07T18:57:29.412708Z","annotations":{"authorization.k8s.io/decision":"allow","authorization.k8s.io/reason":"","pod-security.kubernetes.io/enforce-policy":"privileged:latest"}}
{"kind":"Event","apiVersion":"audit.k8s.io/v1","level":"Metadata","auditID":"19df69d5-109a-4f6f-bfa9-e9171cb0e6f3","stage":"ResponseComplete","requestURI":"/api/v1/namespaces/kube-system/pods/kube-controller-manager-k8s-control-plane-1","verb":"get","user":{"username":"system:node:k8s-control-plane-1","groups":["system:nodes","system:authenticated"]},"sourceIPs":["<IP>"],"userAgent":"kubelet/v1.28.4 (linux/amd64) kubernetes/bae2c62","objectRef":{"resource":"pods","namespace":"kube-system","name":"kube-controller-manager-k8s-control-plane-1","apiVersion":"v1"},"responseStatus":{"metadata":{},"code":200},"requestReceivedTimestamp":"2024-02-07T18:57:32.447451Z","stageTimestamp":"2024-02-07T18:57:32.450013Z","annotations":{"authorization.k8s.io/decision":"allow","authorization.k8s.io/reason":""}}
{"kind":"Event","apiVersion":"audit.k8s.io/v1","level":"Metadata","auditID":"9e0ec506-2733-4dbd-91d3-63f9229cb48c","stage":"ResponseComplete","requestURI":"/api/v1/namespaces/kube-system/pods/kube-controller-manager-k8s-control-plane-1","verb":"get","user":{"username":"system:node:k8s-control-plane-1","groups":["system:nodes","system:authenticated"]},"sourceIPs":["<IP>"],"userAgent":"kubelet/v1.28.4 (linux/amd64) kubernetes/bae2c62","objectRef":{"resource":"pods","namespace":"kube-system","name":"kube-controller-manager-k8s-control-plane-1","apiVersion":"v1"},"responseStatus":{"metadata":{},"code":200},"requestReceivedTimestamp":"2024-02-07T18:57:33.047988Z","stageTimestamp":"2024-02-07T18:57:33.050820Z","annotations":{"authorization.k8s.io/decision":"allow","authorization.k8s.io/reason":""}}
{"kind":"Event","apiVersion":"audit.k8s.io/v1","level":"Metadata","auditID":"1f3da05d-83f5-4112-81c1-0fcd0fe67632","stage":"ResponseComplete","requestURI":"/api/v1/namespaces/kube-system/pods/kube-controller-manager-k8s-control-plane-1","verb":"get","user":{"username":"system:node:k8s-control-plane-1","groups":["system:nodes","system:authenticated"]},"sourceIPs":["<IP>"],"userAgent":"kubelet/v1.28.4 (linux/amd64) kubernetes/bae2c62","objectRef":{"resource":"pods","namespace":"kube-system","name":"kube-controller-manager-k8s-control-plane-1","apiVersion":"v1"},"responseStatus":{"metadata":{},"code":200},"requestReceivedTimestamp":"2024-02-07T18:57:33.152001Z","stageTimestamp":"2024-02-07T18:57:33.154693Z","annotations":{"authorization.k8s.io/decision":"allow","authorization.k8s.io/reason":""}}
```
</details>

Snippet of kubectl output in failed case:
<details>

```console
I0207 18:57:29.319725  114440 request.go:1212] Request Body: {"propagationPolicy":"Background"}
I0207 18:57:29.319811  114440 round_trippers.go:466] curl -v -XDELETE  -H "Accept: application/json" -H "Content-Type: application/json" -H "User-Agent: kubectl/v1.28.4 (linux/amd64) kubernetes/bae2c62" 'https://k8s.example.com:6443/api/v1/namespaces/kube-system/pods/kube-controller-manager-k8s-control-plane-1'
I0207 18:57:29.351133  114440 round_trippers.go:553] DELETE https://k8s.example.com:6443/api/v1/namespaces/kube-system/pods/kube-controller-manager-k8s-control-plane-1 200 OK in 31 milliseconds
I0207 18:57:29.351682  114440 request.go:1212] Response Body: {...}
I0207 18:57:29.352264  114440 round_trippers.go:466] curl -v -XGET  -H "Accept: application/json" -H "User-Agent: kubectl/v1.28.4 (linux/amd64) kubernetes/bae2c62" 'https://k8s.example.com:6443/api/v1/namespaces/kube-system/pods/kube-controller-manager-k8s-control-plane-1'
I0207 18:57:29.367285  114440 round_trippers.go:553] GET https://k8s.example.com:6443/api/v1/namespaces/kube-system/pods/kube-controller-manager-k8s-control-plane-1 200 OK in 14 milliseconds
I0207 18:57:29.370362  114440 request.go:1212] Response Body: {...}
I0207 18:57:29.371341  114440 reflector.go:289] Starting reflector *unstructured.Unstructured (0s) from vendor/k8s.io/client-go/tools/watch/informerwatcher.go:146
I0207 18:57:29.371368  114440 reflector.go:325] Listing and watching *unstructured.Unstructured from vendor/k8s.io/client-go/tools/watch/informerwatcher.go:146
I0207 18:57:29.371502  114440 round_trippers.go:466] curl -v -XGET  -H "Accept: application/json" -H "User-Agent: kubectl/v1.28.4 (linux/amd64) kubernetes/bae2c62" 'https://k8s.example.com:6443/api/v1/namespaces/kube-system/pods?fieldSelector=metadata.name%3Dkube-controller-manager-k8s-control-plane-1&limit=500&resourceVersion=0'
I0207 18:57:29.384838  114440 round_trippers.go:553] GET https://k8s.example.com:6443/api/v1/namespaces/kube-system/pods?fieldSelector=metadata.name%3Dkube-controller-manager-k8s-control-plane-1&limit=500&resourceVersion=0 200 OK in 13 milliseconds
I0207 18:57:29.385198  114440 request.go:1212] Response Body: {"kind":"PodList","apiVersion":"v1","metadata":{"resourceVersion":"84421"},"items":[]}
I0207 18:57:29.385454  114440 round_trippers.go:466] curl -v -XGET  -H "Accept: application/json" -H "User-Agent: kubectl/v1.28.4 (linux/amd64) kubernetes/bae2c62" 'https://k8s.example.com:6443/api/v1/namespaces/kube-system/pods?allowWatchBookmarks=true&fieldSelector=metadata.name%3Dkube-controller-manager-k8s-control-plane-1&resourceVersion=84421&timeoutSeconds=307&watch=true'
I0207 18:57:29.386635  114440 round_trippers.go:553] GET https://k8s.example.com:6443/api/v1/namespaces/kube-system/pods?allowWatchBookmarks=true&fieldSelector=metadata.name%3Dkube-controller-manager-k8s-control-plane-1&resourceVersion=84421&timeoutSeconds=307&watch=true 200 OK in 1 milliseconds
I0207 18:57:29.472007  114440 shared_informer.go:341] caches populated
I0207 19:02:36.387218  114440 reflector.go:790] vendor/k8s.io/client-go/tools/watch/informerwatcher.go:146: Watch close - *unstructured.Unstructured total 10 items received
I0207 19:02:36.387403  114440 round_trippers.go:466] curl -v -XGET  -H "Accept: application/json" -H "User-Agent: kubectl/v1.28.4 (linux/amd64) kubernetes/bae2c62" 'https://k8s.example.com:6443/api/v1/namespaces/kube-system/pods?allowWatchBookmarks=true&fieldSelector=metadata.name%3Dkube-controller-manager-k8s-control-plane-1&resourceVersion=84947&timeoutSeconds=454&watch=true'
I0207 19:02:36.388846  114440 round_trippers.go:553] GET https://k8s.example.com:6443/api/v1/namespaces/kube-system/pods?allowWatchBookmarks=true&fieldSelector=metadata.name%3Dkube-controller-manager-k8s-control-plane-1&resourceVersion=84947&timeoutSeconds=454&watch=true 200 OK in 1 milliseconds
I0207 19:10:10.389943  114440 reflector.go:790] vendor/k8s.io/client-go/tools/watch/informerwatcher.go:146: Watch close - *unstructured.Unstructured total 9 items received
I0207 19:10:10.390130  114440 round_trippers.go:466] curl -v -XGET  -H "Accept: application/json" -H "User-Agent: kubectl/v1.28.4 (linux/amd64) kubernetes/bae2c62" 'https://k8s.example.com:6443/api/v1/namespaces/kube-system/pods?allowWatchBookmarks=true&fieldSelector=metadata.name%3Dkube-controller-manager-k8s-control-plane-1&resourceVersion=85647&timeoutSeconds=478&watch=true'
I0207 19:10:10.391747  114440 round_trippers.go:553] GET https://k8s.example.com:6443/api/v1/namespaces/kube-system/pods?allowWatchBookmarks=true&fieldSelector=metadata.name%3Dkube-controller-manager-k8s-control-plane-1&resourceVersion=85647&timeoutSeconds=478&watch=true 200 OK in 1 milliseconds
```
</details>

Snippet of kubectl output in successful case:
<details>

```console
I0207 19:14:20.389994  127340 request.go:1212] Request Body: {"propagationPolicy":"Background"}
I0207 19:14:20.390087  127340 round_trippers.go:466] curl -v -XDELETE  -H "Accept: application/json" -H "Content-Type: application/json" -H "User-Agent: kubectl/v1.28.4 (linux/amd64) kubernetes/bae2c62" 'https://k8s.example.com:6443/api/v1/namespaces/kube-system/pods/kube-controller-manager-k8s-control-plane-1'
I0207 19:14:20.407828  127340 round_trippers.go:553] DELETE https://k8s.example.com:6443/api/v1/namespaces/kube-system/pods/kube-controller-manager-k8s-control-plane-1 200 OK in 17 milliseconds
I0207 19:14:20.408542  127340 request.go:1212] Response Body: {}
I0207 19:14:20.415560  127340 round_trippers.go:466] curl -v -XGET  -H "Accept: application/json" -H "User-Agent: kubectl/v1.28.4 (linux/amd64) kubernetes/bae2c62" 'https://k8s.example.com:6443/api/v1/namespaces/kube-system/pods/kube-controller-manager-k8s-control-plane-1'
I0207 19:14:20.418975  127340 round_trippers.go:553] GET https://k8s.example.com:6443/api/v1/namespaces/kube-system/pods/kube-controller-manager-k8s-control-plane-1 200 OK in 3 milliseconds
I0207 19:14:20.419298  127340 request.go:1212] Response Body: {...}
I0207 19:14:20.420043  127340 reflector.go:289] Starting reflector *unstructured.Unstructured (0s) from vendor/k8s.io/client-go/tools/watch/informerwatcher.go:146
I0207 19:14:20.420057  127340 reflector.go:325] Listing and watching *unstructured.Unstructured from vendor/k8s.io/client-go/tools/watch/informerwatcher.go:146
I0207 19:14:20.420157  127340 round_trippers.go:466] curl -v -XGET  -H "Accept: application/json" -H "User-Agent: kubectl/v1.28.4 (linux/amd64) kubernetes/bae2c62" 'https://k8s.example.com:6443/api/v1/namespaces/kube-system/pods?fieldSelector=metadata.name%3Dkube-controller-manager-k8s-control-plane-1&limit=500&resourceVersion=0'
I0207 19:14:20.421406  127340 round_trippers.go:553] GET https://k8s.example.com:6443/api/v1/namespaces/kube-system/pods?fieldSelector=metadata.name%3Dkube-controller-manager-k8s-control-plane-1&limit=500&resourceVersion=0 200 OK in 1 milliseconds
I0207 19:14:20.422297  127340 request.go:1212] Response Body: {"kind":"PodList","apiVersion":"v1","metadata":{"resourceVersion":"86043"},"items":[{<one pod>}]}
I0207 19:14:20.423214  127340 round_trippers.go:466] curl -v -XGET  -H "Accept: application/json" -H "User-Agent: kubectl/v1.28.4 (linux/amd64) kubernetes/bae2c62" 'https://k8s.example.com:6443/api/v1/namespaces/kube-system/pods?allowWatchBookmarks=true&fieldSelector=metadata.name%3Dkube-controller-manager-k8s-control-plane-1&resourceVersion=86043&timeoutSeconds=429&watch=true'
I0207 19:14:20.424544  127340 round_trippers.go:553] GET https://k8s.example.com:6443/api/v1/namespaces/kube-system/pods?allowWatchBookmarks=true&fieldSelector=metadata.name%3Dkube-controller-manager-k8s-control-plane-1&resourceVersion=86043&timeoutSeconds=429&watch=true 200 OK in 1 milliseconds
I0207 19:14:20.520374  127340 shared_informer.go:341] caches populated
I0207 19:14:20.520649  127340 reflector.go:295] Stopping reflector *unstructured.Unstructured (0s) from vendor/k8s.io/client-go/tools/watch/informerwatcher.go:146
```
</details>

In comparison to successful case, `GET /api/v1/namespaces/kube-system/pods?fieldSelector=metadata.name%3Dkube-controller-manager-k8s-control-plane-1` returned empty list in failed case.

Some events from audit (timestamps are taken from requestReceivedTimestamp)
18:57:29.329926Z (audit) kubectl deleted pod
18:57:29.344031Z (audit) kubelet deleted pod
18:57:29.353780Z (audit) kubectl queried pod
18:57:29.371784Z (audit) kubectl queried pods by `fieldSelector=metadata.name%3Dkube-controller-manager-k8s-control-plane-1`

Some events from kubectl output.
18:57:29.351133 (kubectl) sent DELETE /api/v1/namespaces/kube-system/pods/kube-controller-manager-k8s-control-plane-1
18:57:29.351682 (kubectl) sent GET /api/v1/namespaces/kube-system/pods/kube-controller-manager-k8s-control-plane-1 and received a pod
18:57:29.384838 (kubectl) sent GET /api/v1/namespaces/kube-system/pods?fieldSelector=metadata.name%3Dkube-controller-manager-k8s-control-plane-1 and received an **empty** list

Probably kubelet somehow affected behaviour of kubectl by parallel deletion of the pod.

### Kubernetes version

<details>

```console
$ kubectl version
Client Version: v1.28.4
Kustomize Version: v5.0.4-0.20230601165947-6ce0bf390ce3
Server Version: v1.28.4
```

</details>

### Cloud provider

<details>
Bare-metal
</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
PRETTY_NAME="Ubuntu 22.04.1 LTS"
NAME="Ubuntu"
VERSION_ID="22.04"
VERSION="22.04.1 LTS (Jammy Jellyfish)"
VERSION_CODENAME=jammy
ID=ubuntu
ID_LIKE=debian
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
UBUNTU_CODENAME=jammy
$ uname -a
Linux k8s-control-plane-1 5.15.0-92-generic #102-Ubuntu SMP Wed Jan 10 09:33:48 UTC 2024 x86_64 x86_64 x86_64 GNU/Linux
```

</details>

### Install tools

<details>
kubeadm
</details>

### Container runtime (CRI) and version (if applicable)

<details>
containerd=1.6.12-0ubuntu1~22.04.3
</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

**Comment by neolit123** [comment]
> Try to delete some static pod kubectl delete pod -n kube-system kube-controller-manager-k8s-control-plane-1

note that these are mirror pods, which serve more of an informal purpose. deleting them will tell the kubelet to re-create them at API level, but this delete action doesn't have much value.

> It seems that the problem is caused by parallel deletion of the pod by kubelet.

your test could be triggering some corner case around static pod sync.
as long as regular pods (non-mirror) can be deleted fine in parallel, without exhibiting this problem, then maybe this could be claimed as a problem around static pods and a "won't fix".

leaving this to sig node for triage, though.
/sig node

if you really want to delete these pods and stop the related containers, you should move/remove the YAML files in /etc/kubernetes/manifests.

**Comment by ilia1243** [comment]
This was a first attempt to do not really delete the pods and containers, but restart the containers and make sure they are restarted and ready. Without pod deletion, immediate attempt to check the pods might show an outdated status. Finally, other solutions was developed avoiding the pod deletion.

**Comment by ndixita** [comment]
cc: @smarterclayton

**Comment by ndixita** [comment]
cc: @harche

**Comment by yujuhong** [comment]
I'm not sure this problem was completely caused by static pod deletions. I'd suggest repeat the steps for static pods that are *not* k8s control plane to isolate the problem to see if it still exists.

**Comment by haircommander** [comment]
Since we're using static pods here, I would recommend you delete the static pod manifests and let the kubelet take care of deleting the pods. I find the pattern of going behind its back to delete with crictl first odd and would recommend against it.

**Comment by ilia1243** [comment]
I need to restart the container, bot do not delete the static pod manifest. Restart is done through crictl. Pod deletion was supposed to refresh the readiness status to wait for (after it is created back automatically). Other ways to restart the container by updating or delete & create the manifest seem approximately the same weird, but I would appreciate if you could officially recommend something.

**Comment by haircommander** [comment]
> delete & create the manifest seem approximately the same weird

I hear you. Often static pod files are managed by some controller or operator so this kind of manual intervention isn't needed. If you'd prefer to manage the lifecycle of the pod, maybe a non static pod would be better. But for static pod management, the most idiomatic way to restart or delete it is to change or delete the manifest.

**Comment by ilia1243** [comment]
Ok, please close the issue if decide to won't fix
